packages/angular/cli/src/utilities/completion.ts

```diff
@@ -0,0 +1,237 @@
+ * so prompts and sets up autocompletion for the user. Returns an exit code if the program should
+ * terminate, otherwise returns `undefined`.
+ * @returns an exit code if the program should terminate, undefined otherwise.
+ */
+export async function considerSettingUpAutocompletion(
+  logger: logging.Logger,
+): Promise<number | undefined> {
+  // Check if we should prompt the user to setup autocompletion.
+  const completionConfig = await getCompletionConfig();
+  if (!(await shouldPromptForAutocompletionSetup(completionConfig))) {
+    return undefined; // Already set up or prompted previously, nothing to do.
+  }
+
+  // Prompt the user and record their response.
+  const shouldSetupAutocompletion = await promptForAutocompletion();
+  if (!shouldSetupAutocompletion) {
+    // User rejected the prompt and doesn't want autocompletion.
+    logger.info(
+      `
+Ok, you won't be prompted again. Should you change your mind, the following command will set up autocompletion for you:
+
+    ${colors.yellow(`ng completion init`)}
+    `.trim(),
+    );
+
+    // Save configuration to remember that the user was prompted and avoid prompting again.
+    await setCompletionConfig({ ...completionConfig, prompted: true });
+
+    return undefined;
+  }
+
+  // User accepted the prompt, set up autocompletion.
+  let rcFile: string;
+  try {
+    rcFile = await initializeAutocomplete();
+  } catch (err) {
+    // Failed to set up autocompeletion, log the error and abort.
+    logger.error(err.message);
+
+    return 1;
+  }
+
+  // Notify the user autocompletion was set up successfully.
+  logger.info(
+    `
+Appended \`source <(ng completion)\` to \`${rcFile}\`. Restart your terminal or run the following to autocomplete \`ng commands\`:
+
+    ${colors.yellow(`source <(ng completion)`)}
+    `.trim(),
+  );
+
+  // Save configuration to remember that the user was prompted.
+  await setCompletionConfig({ ...completionConfig, prompted: true });
+
+  return undefined;
+}
+
+async function getCompletionConfig(): Promise<CompletionConfig | undefined> {
+  const wksp = await getWorkspace('global');
+
+  return wksp?.getCli()?.['completion'];
+}
+
+async function setCompletionConfig(config: CompletionConfig): Promise<void> {
+  const wksp = await getWorkspace('global');
+  if (!wksp) {
+    throw new Error(`Could not find global workspace`);
+  }
+
+  wksp.extensions['cli'] ??= {};
+  const cli = wksp.extensions['cli'];
+  if (!json.isJsonObject(cli)) {
+    throw new Error(
+      `Invalid config found at ${wksp.filePath}. \`extensions.cli\` should be an object.`,
+    );
+  }
+  cli.completion ??= config as json.JsonObject;
+  await wksp.save();
+}
+
+async function shouldPromptForAutocompletionSetup(config?: CompletionConfig): Promise<boolean> {
+  // Force whether or not to prompt for autocomplete to give an easy path for e2e testing to skip.
+  const forceAutocomplete = env['NG_FORCE_AUTOCOMPLETE'];
```

Comment:
Can you please move env variables to packages/angular/cli/src/utilities/environment-options.ts?